### PR TITLE
Issue 30/file paths in config files should be interpreted relative to the file in which they appear

### PR DESCRIPTION
**Pull request by ianmak**
fix for issue #30 
- added support for directory aware configuration loader
- removed —config-path command line argument

### CODE REVIEW

File: user_sync/config.py [hunk 1]
```diff
@@ -60,83 +60,133 @@ def __init__(self, caller_options):
...
-            if (accessor_config_file_path != None):
-                accessor_config_sources.append(accessor_config_file_path)
-            accessor_config_sources.append({            
+            # first, load the accessor configuration file returned from the
+            # filename pattern matcher
+            accessor_config_filename = accessor_config_file_paths.get(organization_name, None)
+
+            # accessor configuration files specified directly in dashboard
+            # accessors takes precedence over the file mapped in the file
+            # pattern
```
Comment (adobeDan) [inline user_sync/config.py]: yeah, this is clear as mud.  please call me to discuss 408-536-4150.

File: user_sync/config.py [hunk 2]
```diff
@@ -158,55 +208,74 @@ def get_directory_connector_options(self, connector_name):
...
+        
         configs = [connector_options, self.options['directory_connector_overridden_options']]
-        current_config = self.combine_dicts(configs)
+        current_config = combine_dicts(configs)
         credential_config_source = credential_manager.get_credentials(credential_manager.DIRECTORY_CREDENTIAL_TYPE, connector_name, config = current_config, config_loader = self)
         return self.get_dict_from_sources([current_config, credential_config_source], "directory_credential_manager")
     
     def get_directory_groups(self):
         '''
+        load the customer group to Adobe dashboard groups mappings model from
```
Comment (adobeDan) [inline user_sync/config.py]: same issue: give overview here, then leave out 1+1=2 comments.

File: user_sync/config.py [hunk 3]
```diff
@@ -158,55 +208,74 @@ def get_directory_connector_options(self, connector_name):
...
+        # load source filters
         source_filters =  self.get_dict_from_sources(source_filter_sources, 'directory[%s].source_filters' % connector_name)
-        if (len(source_filters) > 0):   
+        if source_filters:
             connector_options['source_filters'] = source_filters
-                
+        
         configs = [connector_options, self.options['directory_connector_overridden_options']]
-        current_config = self.combine_dicts(configs)
+        current_config = combine_dicts(configs)
```
Comment (adobeDan) [inline user_sync/config.py]: combine_dicts is not a method?

File: user_sync/config.py [hunk 4]
```diff
@@ -223,25 +292,37 @@ def as_list(value):
         
     def get_dict_from_sources(self, sources, owner):
         '''
+        combine various configuration sources into a single combined dictionary
         :type sources: list
+        :rtype dict
         '''
-        if (sources == None):
+        # sources are empty or None, return an empty dictionary
```
Comment (adobeDan) [inline user_sync/config.py]: really?!  come on.

File: user_sync/config.py [hunk 5]
```diff
@@ -223,25 +292,37 @@ def as_list(value):
...
+        # process sources
         options = []
         for source in sources: 
+            # if the specified source is a string, treat it as a path to a
+            # file containing the source containing a dict
             if (isinstance(source, types.StringTypes)):
                 absolute_path = self.get_absolute_file_path(source)
                 if (os.path.isfile(absolute_path)):
-                    config = self.load_from_yaml(absolute_path)
+                    config = load_from_yaml(absolute_path)
```
Comment (adobeDan) [inline user_sync/config.py]: load_from_yaml is not a method?

File: user_sync/config.py [hunk 6]
```diff
@@ -445,7 +511,7 @@ def check_unused_config_keys(self):
             if (len(unused_keys) > 0):
                 raise user_sync.error.AssertionException("Unused source filters for: %s" % list(unused_keys))
         
-    
+# base class for configuration tree
```
Comment (adobeDan) [inline user_sync/config.py]: not really a helpful comment, but if you are going to have a comment on a class, you should make it the string documentation (first entry in the class).

File: user_sync/config.py [hunk 7]
```diff
@@ -637,6 +727,91 @@ def describe_unused_values(self):
         if (len(unused_keys) > 0):
             messages.append("Found unused keys: %s in: %s" % (unused_keys, self.get_full_scope()))
         return messages
+
+def load_from_yaml(filename):
```
Comment (adobeDan) [inline user_sync/config.py]: moving these out of the class is not really good unless you're trying to export them from the module separately.  it's better to control their visibility through the class.  why did you do this extract?

File: user_sync/config.py [hunk 8]
```diff
@@ -423,15 +476,28 @@ def get_rule_options(self):
         }
         return result
 
-    def create_dashboard_options(self, connector_config_sources, owner):
-        connector_config = self.get_dict_from_sources(connector_config_sources, owner)
-        enterprise_section = connector_config.get('enterprise')
+    def create_dashboard_options(self, config_filename, additional_connector_config_sources, owner):
```
Comment (adobeDan) [inline user_sync/config.py]: Why did you feel the need to distinguish the position of the filename as opposed to letting any entry in the list be a filename and just act as another source?  That's the "YAML way".

File: user_sync/config.py [hunk 9]
```diff
@@ -574,12 +646,17 @@ def iter_unused_keys(self):
         
     def get_dict_config(self, key, none_allowed = False):
         '''
+        returns a 
+        :key: str
         :rtype DictConfig
         '''
+        # check the child configurations for the key
```
Comment (adobeDan) [inline user_sync/config.py]: a lot of 1+1 here.

File: user_sync/config.py [hunk 10]
```diff
@@ -637,6 +727,91 @@ def describe_unused_values(self):
...
+            for dict_key, dict_item in dict_item.iteritems():
+                result_item = result.get(dict_key)
+                if (isinstance(result_item, dict) and isinstance(dict_item, dict)):
+                    result_item.update(dict_item)
+                else:
+                    result[dict_key] = dict_item
+    return result
+
+class ConfigFileLoader(object):
+    def __init__(self, filename):
```
Comment (adobeDan) [inline user_sync/config.py]: You add a whole new class, and there's no class comment? :)

File: user_sync/config.py [hunk 11]
```diff
@@ -637,6 +727,91 @@ def describe_unused_values(self):
...
+        return ConfigFileLoader(self.get_relative_filename(filename))
+        
+    def get_relative_filename(self, filename):
+        '''
+        returns a filename relative to this configuration's directory context
+        :type filename: str
+        :rtype str
+        '''
+        # if the filename is not absolute, process the relative path
+        if not os.path.isabs(filename):
```
Comment (adobeDan) [inline user_sync/config.py]: why is this and the inner if not an and?

File: user_sync/config.py [hunk 12]
```diff
@@ -60,83 +60,133 @@ def __init__(self, caller_options):
             'update_user_info': True,
             'username_filter_regex': None,
         }
-        options.update(caller_options)     
+        
+        # merge custom options with our default options
```
Comment (ianmak) [inline user_sync/config.py]: okay sounds good, I'll try to refactoring the function descriptions and commenting

**Comment by adobeDan** [comment]
Hey, @ianmak, great conversation.  I'm going to close this PR and look forward to you submitting a far more restricted one for tonight based on our discussion.  (You can always come here to look at the review comments, even after the PR is closed.)

**Pull request closed without merging**